lib/iris/common/metadata.py

```diff
@@ -0,0 +1,402 @@
+        dictionary.
+
+        """
+        standard_name = self.standard_name
+        long_name = self.long_name
+        var_name = self.var_name
+
+        # Defensive enforcement of attributes being a dictionary.
+        if not isinstance(self.attributes, Mapping):
+            try:
+                self.attributes = dict()
+            except AttributeError:
+                emsg = "Invalid '{}.attributes' member, must be a mapping."
+                raise AttributeError(emsg.format(self.__class__.__name__))
+
+        stash_name = self.attributes.get("STASH")
+        if stash_name is not None:
+            stash_name = str(stash_name)
+
+        return (standard_name, long_name, var_name, stash_name)
+
+
+def MetadataFactory(cls, **kwargs):
+    """
+    A class instance factory function responsible for manufacturing
+    metadata instances dynamically at runtime.
+
+    The factory instances returned by the factory are capable of managing
+    their metadata state, which can be proxied by the owning container.
+
+    Args:
+
+    * cls:
+        A subclass of :class:`~iris.common.metadata.BaseMetadata`, defining
+        the metadata to be managed.
+
+    Kwargs:
+
+    * kwargs:
+        Initial values for the manufactured metadata instance. Unspecified
+        fields will default to a value of 'None'.
+
+    """
+
+    def __init__(self, cls, **kwargs):
+        # Restrict to only dealing with appropriate metadata classes.
+        if not issubclass(cls, BaseMetadata):
+            emsg = "Require a subclass of {!r}, got {!r}."
+            raise TypeError(emsg.format(BaseMetadata.__name__, cls))
+
+        #: The metadata class to be manufactured by this factory.
+        self.cls = cls
+
+        # Initialise the metadata class fields in the instance.
+        for field in self.fields:
+            setattr(self, field, None)
+
+        # Populate with provided kwargs, which have already been verified
+        # by the factory.
+        for field, value in kwargs.items():
+            setattr(self, field, value)
+
+    def __eq__(self, other):
+        if not hasattr(other, "cls"):
+            return NotImplemented
+        match = self.cls is other.cls
+        if match:
+            match = self.values == other.values
+        return match
+
+    def __getstate__(self):
+        """Return the instance state to be pickled."""
+        return {field: getattr(self, field) for field in self.fields}
+
+    def __ne__(self, other):
+        match = self.__eq__(other)
+        if match is not NotImplemented:
+            match = not match
+        return match
+
+    def __reduce__(self):
+        """
+        Dynamically created classes at runtime cannot be pickled, due to not
+        being defined at the top level of a module. As a result, we require to
+        use the __reduce__ interface to allow 'pickle' to recreate this class
+        instance, and dump and load instance state successfully.
```

Comment:
Reference:
- https://docs.python.org/3/library/pickle.html#what-can-be-pickled-and-unpickled